### PR TITLE
ENH: state space: optional use of Lutkepohl 2010 info. criteria

### PR DESCRIPTION
**Pull request by ChadFulton**
Add the option to specify computing AIC, BIC, and HQIC using Lutkepohl's (2010) alternate formulas.

Notes for consideration / review:

- Right now, the keyword argument (passed to `fit`, `filter`, or `smooth`) is `lutkepohl_stats`, default is False.
- As in the AR / ARIMA / VAR models, I used the true number of parameters (including constant and variance parameters) rather than just the AR / VAR coefficients. This is as opposed to Stata, which ignore the constant and variance parameters in the IC penalty term.
- Unit tests are against Stata and use the SARIMAX and VARMAX models as their examples. In the unit tests I adjust the IC so that the numbers are comparable to Stata.

### CODE REVIEW

**Comment by josef-pkt** [comment]
Why is `lutkepohl_stats` needed as a `fit` keyword? Shouldn't this be a results method?

**Comment by josef-pkt** [comment]
(don't forget to add labels to issues and PRs)

**Comment by ChadFulton** [comment]
> Why is lutkepohl_stats needed as a fit keyword? Shouldn't this be a results method?

It could be. I guess I'm still not sure exactly how the results class should operate.

- The original way I implemented it was as a results property, where the property setter (e.g. `results.lutkepohl_stats = True` reset the cache for aic, bic, hqic variables so that when they were next called they would return the Lutkepohl versions. Unfortunately, it looks like the wrapper doesn't work with properties.
- I guess I could change this to a `use_lutkepohl_stats` method or something.
- Or should I have different attributes? e.g. `aic_lutkepohl`?

I thought I remembered some discussion about covariance matrices, that there was a movement away from e.g. `_get_robustcov_results` returning a whole new results instance? So I thought it was better to specify `lutkepohl_stats` in the construction of the class, which means in `fit`.

But I really have no preference here, happy to implement it whatever way is consistent with Statsmodels.

**Comment by ChadFulton** [comment]
Whoops, didn't include the new results / testing files. Amended git includes them.

**Comment by josef-pkt** [comment]
a few thoughts

In general, post-fit results should be in the results class, but the results instance is created in fit so it's the only place we can add things to `results.__init__`. This was the outcome of the cov_type and use_t decision.
However, I think we should keep those fit keywords to a small number as possible, especially also because it doesn't make it easy to get the non-default result version without calling fit again.

three possibilities:
- add specific keyword like cov_type to fit
- use a more generic keyword in fit like `results_defaults_kwds` requiring a dict (similar to cov_kwds)
- limit the use of defaults and use explicit methods in results

my guess is that information criteria should be in the third case. It doesn't look to me important enough to switch between different ic definitions by default, or is it?
Given that aic, bic are attributes, we could add a separate method for information criteria to the results for different versions, e.g. VAR uses a ic method.

The only place where defaults are really visible is the summary method, and we have some options in those cases, like alpha for the confidence interval.
It might be useful to define a results attribute summary_options if we want to have or give users more flexibility about what is used by defaults (without having to specify a large set of arguments) (I thought there is a related issue for changing what's in the results, but don't find any.)

If the ic versions produce different rankings of models, then they would also affect model selection, which would be a different usage of the defaults than summary.

aside: AIC, BIC in GLM is still inconsistent because they use two different definitions.

**Comment by ChadFulton** [comment]
Adding an `ic` method makes sense to me. So I'll keep the usual `tools.eval_measures` aic, bic, hqic as the attributes, and allow in the `ic` method switching through a `method` keyword argument; with `'standard'` (default) or `'lutkepohl'` as options.

**Comment by ChadFulton** [comment]
Oh, I guess I'll make the method name `info_criteria` for compatibility with `VAR`.

**Comment by ChadFulton** [comment]
> Oh, I guess I'll make the method name info_criteria for compatibility with VAR.

Never mind,`info_criteria` is an attribute in VAR. So, I guess `ic` it is for the method.

**Comment by josef-pkt** [comment]
just in general: the current VAR is not a very good example for consistent informative names (and has quite a bit of refactorings waiting in PRs)

`ic` is a bit short for my taste
I think we could change it in VAR, `info_criteria` sounds like a good method name.

**Comment by ChadFulton** [comment]
Force pushed an amend, which adds a new `info_criteria` method.

**Comment by ChadFulton** [comment]
This looks good to me now, will merge later tonight unless there are concerns.

File: statsmodels/tsa/statespace/mlemodel.py [hunk 1]
```diff
@@ -1885,6 +1891,87 @@ def cov_params_robust_approx(self):
...
+        Information criteria
+
+        Parameters
+        ----------
+        criteria : {'aic', 'bic', 'hqic'}
+            The information criteria to compute.
+        method : {'standard', 'lutkepohl'}
+            The method for information criteria computation. Default is
+            'standard' method; 'lutkepohl' computes the information criteria
+            as in Lutkepohl (2010). See Notes for formulas.
```
Comment (josef-pkt) [inline statsmodels/tsa/statespace/mlemodel.py]: the references have Lutkepohl 2007
?

File: statsmodels/tsa/statespace/mlemodel.py [hunk 2]
```diff
@@ -1885,6 +1891,87 @@ def cov_params_robust_approx(self):
...
+            BIC_L & = \log | Q | + \frac{k \log n}{n} \\
+            HQIC_L & = \log | Q | + \frac{2 k \log \log n}{n} \\
+
+        Note that the Lutkepohl definitions do not apply to all state space
+        models, and should be used with care outside of SARIMAX and VARMAX
+        models.
+
+        References
+        ----------
+        .. [1] Lutkepohl, Helmut. 2007.
```
Comment (josef-pkt) [inline statsmodels/tsa/statespace/mlemodel.py]: can we get the utf-8 correct spelling of the name?

**Comment by josef-pkt** [comment]
looks good to me based on skimming

**Comment by ChadFulton** [comment]
Thanks for the review @josef-pkt

**Pull request merged**